File: ios/NitroGoogleCast/CastButtonRegistry.swift
import GoogleCast
import UIKit

/// Attach-ordered registry of mounted CastButton host views (E10). The single
/// consumer is `showIntroductoryOverlay`, which needs a visible
/// `GCKUICastButton` anchor for GCK's non-deprecated
/// `presentCastInstructionsViewControllerOnce(with:)` API.
///
/// - **Main thread only** (asserted): registration is driven by
///   `didMoveToWindow` and consumed by the transport's main-queue hops.
/// - **Attach order, last attached wins**: `register` (window attach) appends,
///   `unregister` (window detach / recycle) removes, so `current` resolves the
///   most recently attached candidate — mirroring the Android registry.
/// - **Weak entries**: a deallocated host can never be handed out; dead
///   entries are compacted on every mutation.
enum CastButtonRegistry {
  private struct Entry {
    weak var host: CastButtonHostView?
  }

  private static var entries: [Entry] = []

  /// Append `host` as the most recently attached button. Re-registering an
  /// already-known host moves it to the end (it just re-attached — it IS the
  /// last-attached one). Main thread only.
  static func register(_ host: CastButtonHostView) {
    dispatchPrecondition(condition: .onQueue(.main))
    entries.removeAll { $0.host == nil || $0.host === host }
    entries.append(Entry(host: host))
  }

  /// Remove `host` (window detach, recycle). Idempotent. Main thread only.
  static func unregister(_ host: CastButtonHostView) {
    dispatchPrecondition(condition: .onQueue(.main))
    entries.removeAll { $0.host == nil || $0.host === host }
  }

  /// The overlay anchor: the last-attached host that is still in a window and
  /// not hidden (E10 anchor check — attached + `!isHidden`; the hidden check
  /// is on the host because RN applies `style` visibility there), else `nil`.
  /// Main thread only.
  static var current: GCKUICastButton? {
    dispatchPrecondition(condition: .onQueue(.main))
    for entry in entries.reversed() {
      if let host = entry.host, host.window != nil, !host.isHidden {
        return host.castButton
      }
    }
    return nil
  }
}

File: android/src/main/java/com/margelo/nitro/googlecast/CastButtonRegistry.kt
package com.margelo.nitro.googlecast

import android.view.View
import androidx.mediarouter.app.MediaRouteButton
import java.lang.ref.WeakReference

/**
 * Attach-ordered registry of mounted Cast buttons (E10). It exists **only** so
 * `showIntroductoryOverlay` can anchor to "the" CastButton — v5 dialogs no
 * longer need a button. [current] is the most recently attached button that is
 * still attached to a window and `VISIBLE`, mirroring the iOS registry
 * (last-attached wins; attached + not hidden).
 *
 * Weak refs only (a dropped view must never be retained here); entries are
 * added/removed from the attach/detach window callbacks plus `onDropView`.
 * Main-thread only — attach callbacks and the transport's UI methods both run
 * on the main thread.
 */
internal object CastButtonRegistry {
  private val buttons = mutableListOf<WeakReference<MediaRouteButton>>()

  /** Called on attach-to-window. Re-registering moves [button] to the end (last-attached wins). */
  fun register(button: MediaRouteButton) {
    MainThread.assertMainThread("CastButtonRegistry.register")
    remove(button)
    buttons.add(WeakReference(button))
  }

  /** Called on detach-from-window and `onDropView`. Idempotent. */
  fun unregister(button: MediaRouteButton) {
    MainThread.assertMainThread("CastButtonRegistry.unregister")
    remove(button)
  }

  /** The overlay anchor: the last-attached button still attached and visible, else `null`. */
  val current: MediaRouteButton?
    get() {
      MainThread.assertMainThread("CastButtonRegistry.current")
      val anchor =
        buttons.lastOrNull { ref ->
          val button = ref.get()
          button != null && button.isAttachedToWindow && button.visibility == View.VISIBLE
        }
      return anchor?.get()
    }

  /** Drops [button] and any garbage-collected entries. */
  private fun remove(button: MediaRouteButton) {
    buttons.removeAll { it.get() == null || it.get() === button }
  }
}
